### ISSUE TITLE
Using AddSingleton<T>(Func<IServiceProvider, T>) in WebHostBuilder.ConfigureServices with UseStartup<TStartup> cause Func<IServiceProvider, T> to be called twice if TStartup requires T 

### ISSUE THREAD

**Issue by tiakun** [aspnet/Hosting#1411]
Here's a unit test to demonstrate the issue.

```
class MyStartup
{
    public MyStartup(string s) { }
    public void Configure(IApplicationBuilder app) { }
}

[Fact]
public void SingletonTwice()
{
    int count = 0;
    string f()
    {
        count++;
        return Guid.NewGuid().ToString();
    }

    var builder = new WebHostBuilder()
        .ConfigureServices(isp => isp.AddSingleton(i => f()))
        .UseStartup<MyStartup>();

    var server = new TestServer(builder);
    string c2 = server.Host.Services.GetService<string>();
    Assert.Equal(1, count);
}
```

As `f()` is registered as the singleton factory, it should be called only once and `count` should be `1`. The test result is `2`.  

The test use Microsoft.AspNetCore.TestHost 2.0.2, with Microsoft.AspNetCore.Hosting 2.0.2 as a dependency.

The unit test project is as attached:

[SingletonTwice.zip](https://github.com/aspnet/Hosting/files/1977469/SingletonTwice.zip)

**Comment by Tratcher** [comment]
Duplicate of https://github.com/aspnet/Hosting/issues/1150. This is a known issue that won't be fixed in the short term.
